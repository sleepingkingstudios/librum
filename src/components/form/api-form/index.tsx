import * as React from 'react';
import type { FormikValues } from 'formik';
import type { IconDefinition } from '@fortawesome/fontawesome-common-types';

import type {
  Refetch,
  Response,
} from '@api';
import type { Animations } from '@components/types';
import { Form } from '../form';
import type {
  FormStatus,
  OnSubmit,
} from '../types';
import {
  handleSubmit,
  mapStatus,
} from './utils';

export type ApiFormProps = {
  children: React.ReactNode,
  className?: string,
  initialValues: FormikValues,
  loadingAnimation?: Animations,
  loadingIcon?: IconDefinition,
  loadingMessage?: string,
  refetch: Refetch,
  response: Response,
};

export const ApiForm = ({
  refetch,
  response,
  ...rest
}: ApiFormProps): JSX.Element => {
  const { isLoading } = response;
  const status: FormStatus = mapStatus(response);
  const onSubmit: OnSubmit = handleSubmit(refetch);

  return (
    <Form
      initialStatus={status}
      isLoading={isLoading}
      onSubmit={onSubmit}
      {...rest}
    />
  );
};
