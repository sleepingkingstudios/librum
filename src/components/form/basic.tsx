import * as React from 'react';
import {
  Form as FormWrapper,
  Formik,
  FormikValues,
} from 'formik';
import type { OnSubmit } from './types';

interface IBasicFormProps {
  children: React.ReactNode;
  initialValues?: FormikValues;
  onSubmit: OnSubmit;
}

export const BasicForm = (
  { children, initialValues, onSubmit }: IBasicFormProps
): JSX.Element => {
  return (
    <Formik initialValues={initialValues} onSubmit={onSubmit}>
      <FormWrapper>
        { children }
      </FormWrapper>
    </Formik>
  )
};
