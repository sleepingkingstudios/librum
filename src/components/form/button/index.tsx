import * as React from 'react';

import { joinClassNames } from '@utils/react-utils';

interface IFormButtonProps {
  className?: string;
  children: React.ReactNode;
  cols?: (2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12);
  type?: "button" | "reset" | "submit";
}

const columnSpanClassName = ({ cols }: { cols: number }): string | null => {
  if (cols === null) { return null; }

  return `col-span-${cols}`;
};

export const FormButton = ({
  className,
  children,
  cols = null,
  type = 'button',
}: IFormButtonProps): JSX.Element => {
  const joinedClassName = joinClassNames(
    'form-button',
    columnSpanClassName({ cols }),
    className,
  );

  return (
    <button className={joinedClassName} type={type}>{ children }</button>
  );
};
