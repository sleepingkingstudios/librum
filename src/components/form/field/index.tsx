import * as React from 'react';
import {
  capitalize,
  last,
  map,
  snakeCase,
} from 'lodash';
import {
  useField,
  useFormikContext,
} from 'formik';
import type {
  FieldHookConfig,
  FieldInputProps,
} from 'formik';

import type { UpToTwelveColumns } from '@components/types';
import { joinClassNames } from '@utils/react-utils';
import { FormInput } from '../input';
import type { FormStatus } from '../types';
import { getServerErrors } from '../utils';

interface IFormFieldProps {
  className?: string;
  cols?: UpToTwelveColumns;
  id?: string;
  label?: string | false;
  name: string;
  type?: string;
}

interface IFormLabelProps {
  id: string;
  label: string | false;
}

interface IRenderComponentProps {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  field: FieldInputProps<any>;
  id?: string;
  type: string;
}

const columnSpanClassName = ({ cols }: { cols: number }): string | null => {
  if (cols === null) { return null; }

  return `col-span-${cols}`;
};

const convertNameToId = (name: string): string => (
  name.replace(/\]/g, '').split('[').map(snakeCase).join('_')
);

const convertNameToLabel = (name: string): string => {
  const segment: string = last(name.split(/\[/));
  const trimmed: string = snakeCase(segment.replace(/\]/, ''));

  return map(trimmed.split('_'), capitalize).join(' ');
};

const renderComponent = ({
  field,
  id = null,
  type,
}: IRenderComponentProps): JSX.Element => (
  <FormInput field={field} id={id} type={type} />
);

const renderErrors = ({ errors }: { errors: string[] }): JSX.Element => {
  if (errors.length === 0) { return; }

  return (<FormFieldErrors errors={errors} />);
};

const renderLabel = ({ id, label }: IFormLabelProps): JSX.Element => {
  if (label === false) {
    return null;
  }

  return (
    <label className="form-label" htmlFor={id}>{ label }</label>
  );
};

const FormFieldErrors = ({ errors }: { errors: string[] }): JSX.Element => (
  <span className="form-field-errors">{ errors.join(', ') }</span>
);

export const FormField = ({
  className = null,
  cols = null,
  id = null,
  label = null,
  name,
  type = 'text',
}: IFormFieldProps): JSX.Element => {
  const configuredId = id || convertNameToId(name);
  const configuredLabel =
    label === false ? label : (label || convertNameToLabel(name));
  const joinedClassName = joinClassNames(
    'form-field',
    columnSpanClassName({ cols }),
    className,
  );
  const fieldConfig: FieldHookConfig<string> = {
    name,
  };
  const [field] = useField(fieldConfig);
  const context = useFormikContext();
  const status = context.status as FormStatus;
  const errors = getServerErrors({ name, status });

  return (
    <div className={joinedClassName}>
      { renderLabel({ id: configuredId, label: configuredLabel }) }

      { renderComponent({ id: configuredId, field, type }) }

      { renderErrors({ errors }) }
    </div>
  );
};
