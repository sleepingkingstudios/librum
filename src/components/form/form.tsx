import * as React from 'react';
import {
  Form as FormWrapper,
  Formik,
} from 'formik';
import type { IconDefinition } from '@fortawesome/fontawesome-common-types';

import { LoadingOverlay } from '@components/loading-overlay';
import type { Animations } from '@components/types';
import { joinClassNames } from '@utils/react-utils';
import type { FormProps } from './types';

const renderLoadingOverlay = ({
  isLoading,
  loadingAnimation,
  loadingIcon,
  loadingMessage,
}: {
  isLoading: boolean;
  loadingAnimation?: Animations;
  loadingIcon?: IconDefinition;
  loadingMessage?: string;
}): JSX.Element => {
  if (!isLoading) { return null; }

  return (
    <LoadingOverlay
      animate={loadingAnimation}
      icon={loadingIcon}
      message={loadingMessage}
    />
  );
};

export const Form = (
  {
    children,
    className = null,
    initialStatus,
    initialValues,
    isLoading = false,
    loadingAnimation,
    loadingIcon,
    loadingMessage = null,
    onSubmit,
  }: FormProps
): JSX.Element => {
  const joinedClassName = joinClassNames(
    className,
    'relative',
  );

  return (
    <Formik
      initialStatus={initialStatus}
      initialValues={initialValues}
      onSubmit={onSubmit}
    >
      <FormWrapper className={joinedClassName}>
        {
          renderLoadingOverlay({
            isLoading,
            loadingAnimation,
            loadingIcon,
            loadingMessage,
          })
        }

        { children }
      </FormWrapper>
    </Formik>
  );
};
