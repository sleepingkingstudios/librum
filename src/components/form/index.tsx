export { ApiForm } from './api-form';
export { FormButton } from './button';
export { FormField } from './field';
export { Form } from './form';
export { FormInput } from './input';
export { FormRow } from './row';
export { FormSubmitButton } from './submit-button';
