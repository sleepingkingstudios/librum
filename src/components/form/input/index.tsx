import * as React from 'react';
import type { FieldInputProps } from 'formik';

import { joinClassNames } from '@utils/react-utils';

interface IInputProps {
  className?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  field: FieldInputProps<any>;
  id?: string;
  type?: string;
}

export const FormInput = ({
  className,
  field,
  id,
  type = 'text',
}: IInputProps): JSX.Element => {
  const joinedClassName = joinClassNames(
    'form-input',
    className,
  );

  return (
    <input id={id} className={joinedClassName} type={type} {...field} />
  );
};
