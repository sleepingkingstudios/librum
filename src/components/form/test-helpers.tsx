import * as React from 'react';
import {
  Form,
  Formik,
  FormikValues,
} from 'formik';

import type { OnSubmit } from './types';

interface IFormWrapperProps {
  initialValues?: FormikValues;
  onSubmit?: OnSubmit;
  submitButton?: true | false;
}

const SubmitButton = (
  { submitButton }: { submitButton: true | false }
): JSX.Element => {
  if (!submitButton) { return null; }

  return (
    <button type="submit">Submit</button>
  );
};

export const formWrapper = (
  { initialValues, onSubmit, submitButton }: IFormWrapperProps) => {
  const FormWrapper = (
    { children }: { children: React.ReactNode }
  ): JSX.Element => (
    <Formik initialValues={initialValues || {}} onSubmit={onSubmit}>
      <Form>
        { children }

        <SubmitButton submitButton={submitButton} />
      </Form>
    </Formik>
  );

  return FormWrapper;
};
