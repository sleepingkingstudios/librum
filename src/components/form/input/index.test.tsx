import * as React from 'react';
import { Field } from 'formik';

import '@testing-library/jest-dom';
import { render } from '@test-helpers/rendering';

import { FormInput } from './index';
import { formWrapper } from '@components/form/test-helpers';

describe('<FormInput />', () => {
  it('should match the snapshot', () => {
    const { asFragment } = render(
      <Field name="email" component={FormInput} />,
      { wrapper: formWrapper({ onSubmit: jest.fn() }) },
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('when the form has data for the input', () => {
    it('should match the snapshot', () => {
      const initialValues = { email: 'alan.bradley@example.com' };

      const { asFragment } = render(
        <Field name="email" component={FormInput} />,
        { wrapper: formWrapper({ initialValues, onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with className: value', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <Field name="email" className="input-important" component={FormInput} />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with id: value', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <Field id="user-email" name="email" component={FormInput} />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with type: value', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <Field name="email" component={FormInput} type="password" />,
        { wrapper: formWrapper({ onSubmit: jest.fn() }) },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
