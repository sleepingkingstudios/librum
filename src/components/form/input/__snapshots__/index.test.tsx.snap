// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormInput /> should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <input
      class="form-input"
      name="email"
      type="text"
      value=""
    />
  </form>
</DocumentFragment>
`;

exports[`<FormInput /> when the form has data for the input should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <input
      class="form-input"
      name="email"
      type="text"
      value="alan.bradley@example.com"
    />
  </form>
</DocumentFragment>
`;

exports[`<FormInput /> with className: value should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <input
      class="form-input input-important"
      name="email"
      type="text"
      value=""
    />
  </form>
</DocumentFragment>
`;

exports[`<FormInput /> with id: value should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <input
      class="form-input"
      id="user-email"
      name="email"
      type="text"
      value=""
    />
  </form>
</DocumentFragment>
`;

exports[`<FormInput /> with type: value should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <input
      class="form-input"
      name="email"
      type="password"
      value=""
    />
  </form>
</DocumentFragment>
`;
